sources/typescript/src/adapters/kaltura-parallel-upload-request-adapter.ts

```diff
@@ -0,0 +1,240 @@
+import { KalturaRequestOptions } from "../api/kaltura-request-options";
+import { KalturaClientOptions } from "../kaltura-client-options";
+import { CancelableAction } from "../cancelable-action";
+import { KalturaUploadConnectionsManager } from "./kaltura-upload-connections-manager";
+import { KalturaUploadRequestAdapter } from "./kaltura-upload-request-adapter";
+
+interface UploadByParallelChunksData {
+    chunkUploadEnabled: boolean;
+
+    /**
+     * bytes uploaded so far
+     */
+    uploaded: number;
+
+    /**
+     * total number of chunks in the current file
+     */
+    totalChunks: number;
+
+    /**
+     * number of bytes per chunk
+     */
+    chunkSize: number;
+
+    /**
+     * number of chunks done uploading
+     */
+    chunksUploaded: number;
+
+    /**
+     * number of the next chunk to start uploading
+     */
+    nextChunkIndex: number;
+}
+
+export class KalturaParallelUploadRequestAdapter extends KalturaUploadRequestAdapter {
+
+    constructor(clientOptions: KalturaClientOptions, defaultRequestOptions: KalturaRequestOptions) {
+        super(clientOptions, defaultRequestOptions);
+        if (clientOptions.parallelUploadsDisabled) {
+            return;
+        }
+        if (KalturaUploadConnectionsManager.getTotalConnections() === -1) {
+            // initialize manager
+            KalturaUploadConnectionsManager.setTotalConnections(clientOptions.maxConcurrentUploadConnections || 6);
+        }
+    }
+
+    transmit(request: KalturaUploadRequest<any>): CancelableAction<any> {
+        return new CancelableAction((resolve, reject, action) => {
+            const { file} = request.getFileInfo();
+            const chunkSize = this._getChunkSize();
+
+            const data: UploadByParallelChunksData = {
+                chunkUploadEnabled: this._chunkUploadSupported(request),
+                uploaded: 0,
+                totalChunks: Math.ceil(file.size / chunkSize),
+                chunkSize: chunkSize,
+                chunksUploaded: 0,
+                nextChunkIndex: 0,
+            };
+
+            let hasListenerRegistered = false;
+
+            let activeAction: CancelableAction<any>;
+
+            const tryUpload = (waitIfNoConnections = true) => {
+                if (KalturaUploadConnectionsManager.retrieveConnection()) {
+                    activeAction = this._uploadChunk(request, data, data.nextChunkIndex)
+                        .then(handleChunkUploadSuccess, handleChunkUploadError);
+                    data.nextChunkIndex += 1;
+                    return true;
+                }
+                else if (waitIfNoConnections) {
+                    KalturaUploadConnectionsManager.addAvailableConnectionsCallback(handleAvailableConnectionNotification);
+                    hasListenerRegistered = true;
+                }
+                return false;
+            };
+
+            const handleChunkUploadError = reason => {
+                activeAction = null;
```

Comment:
you do not release the connection?
where is the retry initiated?

Reply:
added release connection.
no retries, same as current behavior. we discussed that and decided we will go without for now.